### PR TITLE
PR to resolve #13 - ISO 8601 serialization of types ZonedDateTime and OffsetDateTime

### LINKED ISSUE #13: String conversion should always be ISO-8601 compilant
Class `InstantSerializer` relies on `toString` implementation of `ZonedDateTime`, which is only ISO-8601 if zone and offset are the same, which is not true for example when BST is active in Europe/London timezone (see the [javadocs](http://docs.oracle.com/javase/8/docs/api/java/time/ZonedDateTime.html#toString--)).

`DateTimeFormatter` instances should be used instead to ensure producing ISO compilant output.

### PR DESCRIPTION
**Pull request by zkiss**
A solution proposal for issue #13.

### CODE REVIEW

**Comment by zkiss** [comment]
I have chosen to go for ISO compliant serialization and change the expectations in the tests. In cases where a location specific timezone was expected in the deserialized object, the initial conditions were changed to use fixed offset based `ZoneId` objects. This is because what was stated in the previous comment; ISO compliant date strings do not encode the location of the time zone, only the offset.

This PR now can be merged and contains what I think most people would consider an acceptable ISO compliant behaviour.

**Comment by zkiss** [comment]
@cowtowncoder, @beamerblvd, just pinging you guys to have a look, maybe you've missed it.

**Comment by cowtowncoder** [comment]
I haven't done anything on this, but in absence of concerns, I could merge it for inclusion in 2.6, now that 2.5 is safely out.

**Comment by beamerblvd** [comment]
I'll take a look at this tonight and provide feedback. 

Nick

**Comment by jontejj** [comment]
Did you take a look at it? :)

**Comment by zkiss** [comment]
There was some discussion regarding this PR in the issue: https://github.com/FasterXML/jackson-datatype-jsr310/issues/13

We need to make a decision if we still want to merge this PR. A solution to the issue is desired though, and I am also willing to deliver that, however a decision is needed.

**Comment by cowtowncoder** [comment]
Apologies for my silence here, but I really do not have enough expertise here to have specific opinion.
I can see benefits and challenges at high level.

What I think is needed is to figure out who should own this module, and decide. @beamerblvd wrote this module and would be the best candidate, but I think he has been busy lately.
I will send another note on the mailing list; but from my point, I think it is most important to have someone knowledgeable and active so that we can move forward. I don't want progress to stall due to fear of sometimes making sub-optimal decisions; mistakes can be fixed, implementations changed.

**Comment by zkiss** [comment]
I see two course of actions regarding the code:
* Merge this PR and add a new `SerializationFeature` to include timezone information
* Don't merge this PR, add the new `SerializationFetaure`
Both ways will then require further dev work in this module and the joda module to handle the feature properly.

Merging this PR results in ISO compliant date strings but introduces loss of data and potentially a false sense of safety when people are dealing with `ZonedDateTime` objects - when they modify them they won't follow the timezone rules the original instance had before serialization.

**Comment by cowtowncoder** [comment]
@zkiss quick question: have I already asked for a CLA? (https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf)
I don't think so, apologies if I missed it.

Given that there has been very little feedback, I think I would go ahead and let you choose the path, with just one limitation: I think that the default value for new feature should be such that it does not break existing code. This means that ISO-8601 compatibility forcing should _not_ be the default.

An additional suggestion I have, for path forward, would be to actually add code in Joda (etc) to detect additional timezone information, and simply discard it if it can not be handled. While I support strict handling in most cases, here it seems better to be bit more liberal with what you accept, because date/time handling is already a huge hassle. Adding more strictness seems counter-productive.

So:

1. Yes, for adding the new `SerializationFeature` (FORCE_ISO8601_COMPLIANT?)
2. Make default `false`, which in case of this module should mean "include timezone even at expense of ISO-8601 compatibility"
3. Retrofit applicable modules appropriately.

Does this make sense?

**Comment by cowtowncoder** [comment]
Forgot to mention: CLA should be printed, filled-in & signed and then emailed to `info` at fasterxml dot com.

File: src/main/java/com/fasterxml/jackson/datatype/jsr310/ser/InstantSerializerBase.java [hunk 1]
```diff
@@ -66,7 +77,7 @@ public void serialize(T instant, JsonGenerator generator, SerializerProvider pro
         }
         else
         {
-            generator.writeString(instant.toString());
+            generator.writeString(this.toIsoString.apply(instant));
```
Comment (beamerblvd) [inline src/main/java/com/fasterxml/jackson/datatype/jsr310/ser/InstantSerializerBase.java]: I don't understand why this change is necessary. According to the API documentation, `Instant#toString` already outputs a strictly ISO-compliant string.

Comment (beamerblvd) [inline src/main/java/com/fasterxml/jackson/datatype/jsr310/ser/InstantSerializerBase.java]: Please disregard this. I misunderstood in which file this change was taking place.

~~I don't understand why this change is necessary. According to the API documentation, `Instant#toString` already outputs a strictly ISO-compliant string.~~

File: src/main/java/com/fasterxml/jackson/datatype/jsr310/ser/OffsetDateTimeSerializer.java [hunk 2]
```diff
@@ -8,6 +9,7 @@
 
     protected OffsetDateTimeSerializer() {
         super(OffsetDateTime.class, dt -> dt.toInstant().toEpochMilli(),
-                OffsetDateTime::toEpochSecond, OffsetDateTime::getNano);
+                OffsetDateTime::toEpochSecond, OffsetDateTime::getNano,
+                dt -> DateTimeFormatter.ISO_OFFSET_DATE_TIME.format(dt));
```
Comment (beamerblvd) [inline src/main/java/com/fasterxml/jackson/datatype/jsr310/ser/OffsetDateTimeSerializer.java]: You don't need the lambda here. You can replace:

`dt -> DateTimeFormatter.ISO_OFFSET_DATE_TIME.format(dt)`

With:

`DateTimeFormatter.ISO_OFFSET_DATE_TIME::format`

File: src/main/java/com/fasterxml/jackson/datatype/jsr310/ser/ZonedDateTimeSerializer.java [hunk 3]
```diff
@@ -8,6 +9,8 @@
 
     protected ZonedDateTimeSerializer() {
         super(ZonedDateTime.class, dt -> dt.toInstant().toEpochMilli(),
-                ZonedDateTime::toEpochSecond, ZonedDateTime::getNano);
+                ZonedDateTime::toEpochSecond, ZonedDateTime::getNano,
+                // ISO_ZONED_DATE_TIME is not the ISO format, it is an extension of it
+                dt -> DateTimeFormatter.ISO_OFFSET_DATE_TIME.format(dt));
```
Comment (beamerblvd) [inline src/main/java/com/fasterxml/jackson/datatype/jsr310/ser/ZonedDateTimeSerializer.java]: Shouldn't the lambda here check for a setting to determine whether to use `ISO_OFFSET_DATE_TIME` or `ISO_ZONED_DATE_TIME`? Or does that have to come at a later date after changes are made to the Jackson core?

File: src/main/java/com/fasterxml/jackson/datatype/jsr310/ser/InstantSerializerBase.java [hunk 4]
```diff
@@ -37,13 +38,23 @@
 
     private final ToIntFunction<T> getNanoseconds;
 
+    private final Function<T, String> toIsoString;
+
+    protected InstantSerializerBase(Class<T> supportedType, ToLongFunction<T> getEpochMillis,
+                                    ToLongFunction<T> getEpochSeconds, ToIntFunction<T> getNanoseconds)
+    {
+        this(supportedType, getEpochMillis, getEpochSeconds, getNanoseconds, t -> t.toString());
```
Comment (beamerblvd) [inline src/main/java/com/fasterxml/jackson/datatype/jsr310/ser/InstantSerializerBase.java]: Same as a comment below: s/`t -> t.toString()`/`t::toString`/, compiler to confirm if syntax is correct.

**Comment by beamerblvd** [comment]
Given addressing of my comments, this looks good to me. But I want to better understand how this is supposed to address adding zone information—looks like that's missing.

**Comment by zkiss** [comment]
Thanks everyone for the feedback. Closing the PR because this would be a non-backwards compatible change - based on our discussion this should not be merged. Will implement the `SerialisationFeature` + handling in joda + jsr310. I am also planning on getting back here with more thoughts but just could not find the time yet.

File: src/main/java/com/fasterxml/jackson/datatype/jsr310/ser/ZonedDateTimeSerializer.java (hunk 3 above)
Comment (zkiss) [inline src/main/java/com/fasterxml/jackson/datatype/jsr310/ser/ZonedDateTimeSerializer.java]: Yes, this PR only contained the ISO compatibility changes and introduced loss of time zone information. The idea of the `SerializationFeature` came up later.

**Comment by zkiss** [comment]
@cowtowncoder, @beamerblvd Regarding the way forward:

I fully understand and agree in general with the principle of introducing changes in a backwards compatible manner. However after giving it some thought, in my opinion it is impossible in this case to be backwards compatible. This is because the currently released joda module and jsr310 module have exactly the opposite behaviours.

Joda module currently does not preserve time zone information - it works the same way this module would with this PR merged. My guess is that joda module is used much more widely than this one.

For this reason I believe that the default value of the 'time zone preserving' config switch should be to **not preserve** the time zone information and produce iso8601 compliant strings.

*If* we agree in this, then my suggestion would be merging this PR and then adding the switch afterwards. If we don't agree, then the introduction of the new `SerializationFeature` will break backwards compatibility in the joda module, as it will start printing timezone information if someone upgrades to the version which contains the feature (This might not break server-to-server communication where both servers upgrade as the module would handle the presence of the unwanted time zone data, but old modules will fail because of the new format).

It is also a possibility that the joda module would ignore the new `SerializationFeature`, however I would not like that - I'd very much like the joda module and this one to be compatible with each other and behave the same way with the same `ObjectMapper` configuration.

Regarding the name of the `SerializationFeature`, I think it should be explicit about the fact that it is preserving timezone information. Something like `WRITE_DATES_WITH_TIMEZONE` or `PRESERVE_ZONE`. In my opinion being compliant with iso8601 by default should not be a question, as that is what the library has has been so good at so far.

**Comment by cowtowncoder** [comment]
Ok. Yes, I see the conundrum here. I also agree in that changing behavior of Joda module in backwards incompatible way would not be well received.

**Comment by zkiss** [comment]
Reopening PR to enable merging if people agree.

**Comment by cowtowncoder** [comment]
I'll let @beamerblvd comment on this. If I do not hear objections, I can merge this weekend (feb 28 2015)

**Comment by lukasniemeier-zalando** [comment]
Any news on this PR? IMO the change is a good one

**Comment by zkiss** [comment]
@cowtowncoder, @beamerblvd, do you guys still want to merge this?

**Comment by wselwood** [comment]
Any news on this getting merged? 

We've been running a local build of this for a while and all seems good to us.

**Comment by cowtowncoder** [comment]
I do have CLA, so technically this could be merged.

**Comment by cowtowncoder** [comment]
@wselwood Apologies for slowness here, and thank you for keeping the thread active. Sometimes a ping is helpful. :)

@zkiss It looks like this changes behavior, but there is no switch to choose between the new ISO-8601 compliant mode, and older representation that includes timezone (and is round-trippable?).
I thought discussion suggested that such a setting would make this non-controversial; but leaving it out would risk backwards compatibility issues.
Apologies if I misunderstood above.

Assuming we need a switch, I see two possibilities.

1. Add a setting directly in `JodaModule`, locally. It would be easy to set, but not dynamically changeable.
2. Add a new `SerializationFeature` in `jackson-databind`; perhaps something like `FORCE_ISO_8601_COMPLIANT` or such, which would be used for this and similar cases

I think starting with (1) might be best first, and then we could see if a general feature would be useful, that is, are there other similar cases to support, and if so, can they share the setting.

@beamerblvd @tommack Let me know if (when...) I misunderstood something above. I am ready to merge this in, as long as there is a switch to use.

**Comment by zkiss** [comment]
@cowtowncoder @beamerblvd @tommack This PR is intentionally not backwards compatible. This is because the joda module and this module currently have exactly the opposite behaviours; this change adjusts the behaviour of this module to the joda module (more on this: https://github.com/FasterXML/jackson-datatype-jsr310/pull/15#issuecomment-75154854).

The plan (mine, at least) is to introduce a completely new `SerializationFeature` called `WRITE_DATES_WITH_TIMEZONE` *turned off* by default (current behaviour of the joda module). This is the second option @cowtowncoder mentioned in his comment (https://github.com/FasterXML/jackson-datatype-jsr310/pull/15#issuecomment-102115034)

### Possible solutions on the short term (in order of preference)

1. Don't solve it. I can't emphasize enough that the joda module has no such feature; it *does not* preserve timezone information, so whoever needed this had to come up with an in-house solution to this issue (probably sending through the timezone name in a separate attribute or serializing the timezone name in a similar way java 8 does). We'll address this issue later properly by introducing the feature.

2. We could write a new serializer which writes the timezone information - just like the old one did. This would have to be manually registered on the `JSR310Module` to the type `ZonedDateTime` which would overwrite the default one.

3. A `writeDatesWithTimezone(boolean)` (or similar) method could be introduced on the `JSR310Module` class (instantly marked as `@Deprecated` as it will be removed as soon as the feature is introduced). I have a feeling doing this would lead to a bit ugly code.

The downside of solutions 2 and 3 is that the only way to control the behaviour of the module is by having referential access to the module itself. People using jackson through Spring for example would have a hard time extracting the registered instance of the `JSR310Module` from the `ObjectMapper` (I believe it is impossible - they would have to create their own `ObjectMapper` instead).

### Summary

- I think this PR should be merged
  - Maybe support the old behaviour temporarily
  - Removal of such support will cause compilation errors in user code when the new feature is introduced
- The zone preserving feature should be separately introduced later

**Comment by tommack** [comment]
My thoughts:

 - Sacrificing backwards compatibility here to align with the Joda module seems like an acceptable choice today.
 - Having a feature switch honored by both Joda and this module would be great.
 - Removing functionality here with no immediate option to turn it back on is *not* so great.
 - Given a choice between FORCE_ISO_8601_COMPLIANT and WRITE_DATES_WITH_TIMEZONE and would choose the latter.
 - A couple years from now, Joda usage will decline and we'll be living in a world where developers are much more familiar with ZoneId vs ZoneOffset and ZonedDateTime vs OffsetDateTime. Correct serialization will be important.

**Comment by wselwood** [comment]
To stick my nose in here I would much prefer the default to be writing out things that can be easily read by other systems, the standard ISO8601 format.

A switch to include the Zone Id and make things round trip able would be good. But I do have to disagree with the switch name. WRITE_DATES_WITH_TIMEZONE being false implies that you wouldn't get the zone offset ( +0100 ) bit either. WRITE_DATES_WITH_ZONE_ID or some thing similar would be clearer, if less portable, not sure how Joda expresses this.

**Comment by zkiss** [comment]
@wselwood I like `WRITE_DATES_WITH_ZONE_ID` as well

**Comment by beamerblvd** [comment]
I, also, prefer `WRITE_DATES_WITH_ZONE_ID`. But I have strong objections to breaking backwards compatibility in this module. I'm not sure what a good alternative is.

@tommack said:
> A couple years from now, Joda usage will decline and we'll be living in a world where developers are much more familiar with ZoneId vs ZoneOffset and ZonedDateTime vs OffsetDateTime. Correct serialization will be important.

Maybe this is the solution. Eventually (it'll be a while, yes, but eventually) the Joda module will be irrelevant. Maybe we have a switch that _only_ affects this module and doesn't affect Joda. Yes, I think that would be my preferred solution.

Sorry for the train-of-thought comment.

**Comment by cowtowncoder** [comment]
I concur with comments so far: I do not like backwards-incompatibility, and if a choice is needed, I would rather upgrade this module first, to allow transition, and postpone decision on Joda (if need be) until 2.7.

But as one minor starting point, since suggestion to add `SerializationFeature.WRITE_DATES_WITH_ZONE_ID` seems non-controversial, I added issue for adding it:

https://github.com/FasterXML/jackson-databind/issues/794

and can easily add it. From discussions, it seems we could choose default either way; however, it sounds like preferred value would be `false` (in absence of compatibility constraints), so that the base behavior would be to start with ISO-8601 compatibility. If that is the eventual state we would want to get to, I would want to start with it. Otherwise we would have to change that default setting at a later point.

At this point it may be relevant to check out usage. From Maven Central, actual download counts for datatype modules are as follows:

**Comment by zkiss** [comment]
Sounds like we need to provide backwards compatibility options then.

My preferred solution for that would be writing a new serializer `ZonedDateTimeWithZoneIdSerializer` (marked as deprecated? But definitely explained in the javadocs that it is planned to be removed). The module would stop serializing zone ids by default. Users who need the old behaviour would have to do something like this:

```
ObjectMapper om = ... // get reference to OM
JSR310Module m = new JSR310Module();
m.addSerializer(ZonedDateTime.class, ZonedDateTimeWithZoneIdSerializer.INSTANCE);
om.registerModule(m);
```

Is that acceptable to everyone?

**Comment by cowtowncoder** [comment]
@zkiss Unfortunately I am not sure that is any better than other options -- many users simply upgrade versions of modules with no other changes, and expect things to work.
I think the worst kind of problem is one where compilation does not fail, and startup does not fail. but somewhere down the line something is different. This is what would happen here.

One other possibility is that we could remove no-arguments constructor of module, and require it to be constructed with a boolean flag (or similar). This would at least indicate the problem immediately as a linkage error. An alternative would be to keep no-arg constructor but make it throw an exception. This could give a more meaningful failure message.

And why take `boolean`? To indicate whether to force use of old defaults ('true', perhaps) or not (`false`). With 2.7 we could make no-arguments constructor valid for use again.

I am not crazy about this option either, but we don't really have good solutions to choose from. So trying to suggest less-bad (in some aspect) options.

**Comment by cowtowncoder** [comment]
And one more. There is also one variant of the idea above: instead of adding new constructor, we could instead create a new module class, so that:

- Existing `JSR310Module` would default to existing behavior, but marked as deprecated
- New module, name to be determined (`NewJSR310Module?` ;-) ), would use the new configuration

we could easily refactor things so that most code would be shared.

This might actually be the better way to do this, mostly because creating instances using no-args constructors is bit easier with some frameworks. Plus we wouldn't need to consider whether to change no-args constructor of the existing module: we would instead deprecate and (eventually) remove that module class.

**Comment by tommack** [comment]
I think I like the new module idea. How about``Java8DateModule`` or ``CompatibleJSR310Module``.

**Comment by zkiss** [comment]
OK, let's go with the new module idea! I'd mark `JSR310Module` as `@Deprecated` and create a new module which serializes in ISO 8601 compatible way - and pretty soon honours the new `SerializationFeature` as well. As for the name, how about `JavaTimeModule` (making a reference to the package name)?

**Comment by wselwood** [comment]
I'd rather use the `SerializationFeature.WRITE_DATES_WITH_ZONE_ID` feature @cowtowncoder added if possible.

Examples using spring Jackson2ObjectMapperBuilder

```
builder.featuresToEnable(SerializationFeature.WRITE_DATES_WITH_ZONE_ID);
```

over 
```java
JSR310Module m = new JSR310Module();
m.addSerializer(ZonedDateTime.class, ZonedDateTimeWithZoneIdSerializer.INSTANCE);
builder.modules(m);
```

Enabling and disabling other more complex features (ADJUST_DATES_TO_CONTEXT_TIME_ZONE) is as simple as setting the feature. I think the default for that serialization feature was set to disabled so it will usually do things the way we want. It can be enabled if people need it and it gives a much simpler interface. 

I know you ( @zkiss ) would rather people didn't use it but given the usage numbers it has to be easy for people to fix code this will break.

**Comment by sourcekick** [comment]
+1 for new Module, deprecating the old Module. This makes things very clear to framework users.

**Comment by cowtowncoder** [comment]
@wselwood findAndRegisterModules() is an interesting part. If I remember correctly, it uses this file:

    src/main/resources/META-INF/services/com.fasterxml.jackson.databind.Module

and we thereby control what is visible. However, our choice is pretty much just including one, both or none there. This is why I personally never use auto-detection; it tends to lead to accidental inclusion of things due to things getting added to classpath, often accidentally via transitive dependencies.

So the question here would be whether to include old or new module in classpath (I don't think "both" makes sense).
One possibility would be to include old one for 2.6, new for 2.7. But fundamentally it is just optimizing on point of breakage....

The nuclear option would be to create a new project. But that does seem... bit extreme.

**Comment by cowtowncoder** [comment]
As to name: I like both `Java8DateModule` and `JavaTimeModule` -- either one would work for me.

**Comment by wselwood** [comment]
By the sounds of it the best we can do in that case is optimize the point of breaking @cowtowncoder 

My vote would go to JavaTimeModule but I am pretty happy with either name.

**Comment by cowtowncoder** [comment]
@wselwood Agreed. Either name is fine -- what is needed would be updates to PR, to allow both options.
Module name is trivial to change. I will be happy to merge version that does what we have discussed; a backwards-compatible version of existing module, and the new one that uses the new `SerializationFeature.WRITE_DATES_WITH_ZONE_ID` (which defaults to false).

And one more note on auto-registration, this: 

https://github.com/FasterXML/jackson-jaxrs-providers/issues/39

is a fix to JAX-RS provider, which now provides two sets of jars: one with auto-registration metadata; other without one. Maven classifier is used to specify one without metadata, to retain backwards-compatibility.
If it makes sense, we could do the same here if we want to.
But as with naming, this can be done after getting the base functionality in place.

**Comment by cowtowncoder** [comment]
@zkiss Do you think more information is needed here? Can I help in getting this implemented, as discussed? I am hoping to get this in 2.6 (ideally, 0-rc2).

**Comment by zkiss** [comment]
@cowtowncoder and all, thanks for all the input. I have pushed a new module class `JavaTimeModule` which should be doing what we discussed. I have duplicated the tests and explicitly copied the old ZonedDateTime test to check it is not broken.

The module is not autodetected by default. If there are any changes are needed, please let me know!

**Comment by cowtowncoder** [comment]
@zkiss I trust that changes are good, but just in case others have comments, I'll wait until merging this. Hoping to merge tonight or tomorrow.

**Comment by zkiss** [comment]
One thing I forgot to mention: because of the duplication of all the tests (thinking in the short term here), this should either be merged real soon (before merging anything else with test changes) or a new way of testing the old and new behaviour should be implemented in this PR.

**Comment by cowtowncoder** [comment]
I'll postpone other merges, if any (none pending). Thank you for noting this problem. I am fine with duplication on short term.

**Comment by zkiss** [comment]
@cowtowncoder I think the old module `JSR310Module` should be deprecated but I could not make it deprecated because I get an error when doing `mvn clean verify`:
```
[WARNING] /home/zoltan/work/jackson-datatype-jsr310/src/test/java/com/fasterxml/jackson/datatype/jsr310/old/TestYearKeySerialization.java:[9,45] com.fasterxml.jackson.datatype.jsr310.JSR310Module in com.fasterxml.jackson.datatype.jsr310 has been deprecated
```

**Comment by cowtowncoder** [comment]
@zkiss I can deal with that one. Module sets up somewhat strict limits so that no warnings are allowed -- bit extreme, and perhaps could be reconsidered. But for now, I can just fix that, after merge.

**Comment by jontejj** [comment]
As this has been discussed quite much I think many of us are interested in a release. Any possibility to do it in the near future?

**Comment by cowtowncoder** [comment]
This will be included in 2.6.0-rc2, once I resolve couple of issues from `jackson-databind` (not related to this issue). Should be within a week.
